### PR TITLE
nosetests test added for bugzilla utils

### PR DESCRIPTION
**Pull request by anubnair**

### CODE REVIEW

**Comment by sylvestre** [comment]
I guess you saw that the test is failing:
https://travis-ci.org/mozilla/relman-auto-nag/builds/87835701

**Comment by anubnair** [comment]
@sylvestre : The fix has been done. I couldn't reproduce the issue when I was testing in my environment. So I have added "--nocapture" option also in tox.ini, so that the print messages can also capture.

**Comment by sylvestre** [comment]
not sure I understand the action of --nocapture ? Do you have some docs?

**Comment by anoopvalluthadam** [comment]
--nocapture used to capture stdout

normally it captures stdout only when there is an ERROR

I think she used it for testing and forgot to remove!

**Comment by anubnair** [comment]
@anoopvalluthadam Yeah I have added for testing. But I thought its better to keep. @sylvestre : can we keep --nocapture also?

**Comment by sylvestre** [comment]
hmm, seems like --no-capture is going to get less potential information. More info is better, don't you think ? :)

**Comment by anubnair** [comment]
@sylvestre:

Please find the below document:
http://nose.readthedocs.org/en/latest/plugins/capture.html

-s, --nocapture

    Don’t capture stdout (any stdout output will be printed immediately) [NOSE_NOCAPTURE]

if we add --nocapture, all stdout output will be printed immediately, So all the "print" statements will get printed. We can understand the flow, if an error occurs, easily we knows where it  breaks.

File: auto_nag/tests/test_utils.py [hunk 1]
```diff
@@ -37,3 +41,54 @@ def test_hide_personal_info(self):
...
+        child.sendline('test_username')
+        child.expect('password')
+        child.sendline('test_password')
+        child.expect(pexpect.EOF)
+        child.close()
+
+        credentials = get_credentials()
+
+        # Removing the ~/.bztoolsrc which generated for testing.
+        cmd = 'rm ' + rcfile
```
Comment (sylvestre) [inline auto_nag/tests/test_utils.py]: Please use the python api for this.

File: auto_nag/tests/test_utils.py [hunk 2]
```diff
@@ -37,3 +41,54 @@ def test_hide_personal_info(self):
...
+        os.system(cmd)
+
+        # copy back bztoolsrc_backup to ~/.bztoolsrc
+        rcfile_backup = os.path.expanduser('~/.bztoolsrc_backup')
+        if os.path.exists(rcfile_backup):
+            cmd = 'mv ' + rcfile_backup + ' ' + expand_usr + '/.bztoolsrc'
```
Comment (sylvestre) [inline auto_nag/tests/test_utils.py]: ditto, please use the python api. os.system should be avoid at all cost.

File: auto_nag/tests/test_utils.py [hunk 3]
```diff
@@ -37,3 +41,54 @@ def test_hide_personal_info(self):
...
+            os.system(cmd)
+        assert credentials == ('test_username', 'test_password')
```
Comment (sylvestre) [inline auto_nag/tests/test_utils.py]: please use assertEquals

File: tox.ini [hunk 4]
```diff
@@ -12,4 +12,4 @@ deps =
 
 commands =
     coverage erase
-    coverage run --branch --source {toxinidir}/auto_nag {envbindir}/nosetests -v --rednose --force-color {toxinidir}/auto_nag
+    coverage run --branch --source {toxinidir}/auto_nag {envbindir}/nosetests -v --rednose --nocapture --force-color {toxinidir}/auto_nag
```
Comment (sylvestre) [inline tox.ini]: please make this change in a separate commits (or better, PR).
Now, this change in a commit called "Minor fixes".

File: auto_nag/tests/test_utils.py [hunk 5]
```diff
@@ -37,3 +41,54 @@ def test_hide_personal_info(self):
             "include_fields=_default,attachments&changed_field_to=" + \
             "RESOLVED&api_key=xyzxyzxyz&resolution=FIXED"
         assert "xyzxyzxyz" not in hide_personal_info(sample_exception)
+
+    def test_urljoin(self):
+        api_root = 'https://bugzilla.mozilla.org/bzapi/'
```
Comment (sylvestre) [inline auto_nag/tests/test_utils.py]: please move http://bugzilla.mozilla.org into a variable and reuse it (I don't like string duplication).

File: auto_nag/tests/test_utils.py [hunk 6]
```diff
@@ -37,3 +41,54 @@ def test_hide_personal_info(self):
...
+        _qs = qs(**param)
+        expected_qs = '&product=Core,Firefox'
+        assert _qs == expected_qs
+
+    def test_get_credentials(self):
+        # if already bztoolsrc is present, backup the existing '~/.bztoolsrc'
+        expand_usr = os.path.expanduser('~')
+        rcfile = os.path.expanduser('~/.bztoolsrc')
+        if os.path.exists(rcfile):
+            cmd = 'mv ' + rcfile + ' ' + expand_usr + '/.bztoolsrc_backup'
```
Comment (sylvestre) [inline auto_nag/tests/test_utils.py]: please use the python api

File: auto_nag/tests/test_utils.py [hunk 7]
```diff
@@ -37,3 +41,54 @@ def test_hide_personal_info(self):
...
+            os.system(cmd)
+        cmd = 'python ' + get_project_root_path() + \
+            'auto_nag/tests/_get_credential.py'
+        child = pexpect.spawn(cmd, timeout=180)
```
Comment (sylvestre) [inline auto_nag/tests/test_utils.py]: Why do you need to use pexpect here? we could just call the python code normally ?

Comment (anubnair) [inline auto_nag/tests/test_utils.py]: @sylvestre : It takes, user input. To automate raw_input and password, I used pexpect

  username = raw_input('Bugzilla username: ')
  password = getpass.getpass('Bugzilla password: ')

Comment (sylvestre) [inline auto_nag/tests/test_utils.py]: OK, make sense, however, not sure I care about this use case of relman auto nag (sorry).
We should probably remove this use case in the code.

Comment (anubnair) [inline auto_nag/tests/test_utils.py]: @sylvestre :-> sure. get_credential() is used only in attach.py, when we start writing test case for that, the needed changes can be done. For now, let it be like this.

File: tox.ini (hunk 4 above)
Comment (anubnair) [inline tox.ini]: Done

File: auto_nag/tests/test_utils.py (hunk 5 above)
Comment (anubnair) [inline auto_nag/tests/test_utils.py]: @sylvestre Most places we are repeating the urls. So, I am planning to create one more file named settings(inside bugzilla directory), so that we can add all urls, or any other configuration parameters into it.

**Comment by anubnair** [comment]
@sylvestre all changes are done

File: auto_nag/tests/test_utils.py (hunk 5 above)
Comment (sylvestre) [inline auto_nag/tests/test_utils.py]: Great idea!

File: auto_nag/bugzilla/settings.py [hunk 8]
```diff
@@ -0,0 +1,2 @@
+BZ_API_ROOT = "https://bugzilla.mozilla.org/bzapi/"
+API_ROOT = "https://bugzilla.mozilla.org/"
```
Comment (sylvestre) [inline auto_nag/bugzilla/settings.py]: BZ_API_ROOT  = API_ROOT +  "bzapi/" :)

File: auto_nag/tests/test_utils.py (hunk 6 above)
Comment (sylvestre) [inline auto_nag/tests/test_utils.py]: Still the same issue here?!

File: auto_nag/tests/test_utils.py (hunk 2 above)
Comment (sylvestre) [inline auto_nag/tests/test_utils.py]: ditto

File: auto_nag/tests/test_utils.py (hunk 6 above)
Comment (anubnair) [inline auto_nag/tests/test_utils.py]: @sylvestre It changed, 67th line is now using, execute_command() function, which does the following

+def execute_command(cmd):
+    """
+    Execute the given cmd
+    """
+    cmd = shlex.split(cmd)
+    p = subprocess.Popen(cmd, stdout=subprocess.PIPE, close_fds=True)
+    out, err = p.communicate()
+    return (out, err)
+

File: auto_nag/tests/test_utils.py (hunk 2 above)
Comment (anubnair) [inline auto_nag/tests/test_utils.py]: os.system is already removes, cmd variable is still ther, added a common function in utils, execute_command(cmd) will do the job

File: auto_nag/bugzilla/settings.py (hunk 8 above)
Comment (anubnair) [inline auto_nag/bugzilla/settings.py]: Done! :)

File: auto_nag/tests/test_utils.py (hunk 2 above)
Comment (sylvestre) [inline auto_nag/tests/test_utils.py]: https://docs.python.org/2/library/shutil.html
please use shutil.move

I won't acces any PR containing call to system command when alternatives are available in the language.

On the global level, doing this is a bad programming habit because:
* mv could be aliased on the system and have a diff system (alias mv="mv -i" for example)
* mv might not exists on the system (Windows?)
* if it fails (wrong permissions), it is going to be hard to catch the correct exception and manage the error.

**Comment by sylvestre** [comment]
@anubnair "minor changes" => please avoid this kind of title for a commit :)

**Comment by anubnair** [comment]
@sylvestre All changes are done
